File: backend/app/schemas/endereco.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


# Schema para receber os parâmetros de geração em massa
class EnderecoLoteCriar(BaseModel):
    area_id: int
    rua_inicio: int
    rua_fim: int
    predio_inicio: int
    predio_fim: int
    nivel_inicio: int
    nivel_fim: int
    posicao_inicio: int
    posicao_fim: int

    estrutura_fisica_id: int
    finalidade_id: int
    peso_maximo_kg: float

    # Exclusivo para Picking Fixo
    produto_id: Optional[int] = None
    capacidade_maxima_und: Optional[int] = None

    # Status inicial
    ativo: bool = True
    bloqueado: bool = False
    motivo_bloqueio: Optional[str] = None


# Schema para atualização individual de um endereço
class EnderecoAtualizar(BaseModel):
    estrutura_fisica_id: Optional[int] = None
    finalidade_id: Optional[int] = None
    peso_maximo_kg: Optional[float] = None
    produto_id: Optional[int] = None
    capacidade_maxima_und: Optional[int] = None
    ativo: Optional[bool] = None
    bloqueado: Optional[bool] = None
    motivo_bloqueio: Optional[str] = None
    rowversion: int  # Obrigatório para controle de concorrência


# Schema para devolver os dados na leitura
class EnderecoSchema(BaseModel):
    id: int
    area_id: int
    rua: int
    predio: int
    nivel: int
    posicao: int
    codigo_formatado: str
    estrutura_fisica_id: int
    finalidade_id: int
    peso_maximo_kg: float
    produto_id: Optional[int] = None
    capacidade_maxima_und: Optional[int] = None
    ativo: bool
    bloqueado: bool
    motivo_bloqueio: Optional[str] = None
    criado_em: datetime
    atualizado_em: datetime
    rowversion: int

    class Config:
        from_attributes = True


# Schema detalhado que inclui os nomes das tabelas relacionadas
class EnderecoDetalhadoSchema(BaseModel):
    id: int
    area_id: int
    rua: int
    predio: int
    nivel: int
    posicao: int
    codigo_formatado: str
    estrutura_fisica_id: int
    finalidade_id: int
    peso_maximo_kg: float
    produto_id: Optional[int] = None
    capacidade_maxima_und: Optional[int] = None
    ativo: bool
    bloqueado: bool
    motivo_bloqueio: Optional[str] = None
    criado_em: datetime
    atualizado_em: datetime
    rowversion: int

    # Campos expandidos (nomes legíveis)
    area_letra: Optional[str] = None
    estrutura_nome: Optional[str] = None
    finalidade_nome: Optional[str] = None
    produto_descricao: Optional[str] = None

    class Config:
        from_attributes = True

File: frontend/src/types/endereco.ts
// --- Endereço ---

export interface Endereco {
  id: number;
  area_id: number;
  rua: number;
  predio: number;
  nivel: number;
  posicao: number;
  codigo_formatado: string;
  estrutura_fisica_id: number;
  finalidade_id: number;
  peso_maximo_kg: number;
  produto_id: number | null;
  capacidade_maxima_und: number | null;
  ativo: boolean;
  bloqueado: boolean;
  motivo_bloqueio: string | null;
  criado_em: string;
  atualizado_em: string;
  rowversion: number;

  // Campos expandidos
  area_letra: string | null;
  estrutura_nome: string | null;
  finalidade_nome: string | null;
  produto_descricao: string | null;
}

export interface EnderecoLoteCriar {
  area_id: number;
  rua_inicio: number;
  rua_fim: number;
  predio_inicio: number;
  predio_fim: number;
  nivel_inicio: number;
  nivel_fim: number;
  posicao_inicio: number;
  posicao_fim: number;
  estrutura_fisica_id: number;
  finalidade_id: number;
  peso_maximo_kg: number;
  produto_id?: number | null;
  capacidade_maxima_und?: number | null;
  ativo?: boolean;
  bloqueado?: boolean;
  motivo_bloqueio?: string | null;
}

export interface EnderecoAtualizar {
  estrutura_fisica_id?: number;
  finalidade_id?: number;
  peso_maximo_kg?: number;
  produto_id?: number | null;
  capacidade_maxima_und?: number | null;
  ativo?: boolean;
  bloqueado?: boolean;
  motivo_bloqueio?: string | null;
  rowversion: number;
}

// --- Tabelas de Apoio ---

export interface Area {
  id: number;
  letra: string;
  descricao: string;
  filial_id: number;
}

export interface EstruturaFisica {
  id: number;
  nome: string;
  comporta_palete: boolean;
  comporta_caixa: boolean;
  comporta_log: boolean;
}

export interface FinalidadeEndereco {
  id: number;
  nome: string;
  tipo_pulmao: boolean;
  tipo_picking: boolean;
  tipo_quarentena: boolean;
}

// --- Produto (simplificado para selecionar no endereço) ---

export interface ProdutoSimples {
  id: number;
  sku: string;
  descricao: string;
}
